File: 4.8-bc/py/stage-1-flat/app/service/notification.py
from __future__ import annotations

import logging
from datetime import UTC, datetime
from uuid import UUID, uuid4

from app.model.notification import Notification
from app.repository.notification import NotificationRepo

logger = logging.getLogger(__name__)


class NotificationService:
    def __init__(self, repo: NotificationRepo) -> None:
        self._repo = repo

    async def send_test(
        self, user_id: UUID, channel: str, payload: str
    ) -> Notification:
        now = datetime.now(UTC)
        n = Notification(
            id=uuid4(),
            user_id=user_id,
            channel=channel,
            payload=payload,
            sent_at=now,
        )
        logger.info("[stub-sender] %s -> user=%s payload=%r", channel, user_id, payload)
        await self._repo.create(n)
        return n

File: 4.8-bc/go/stage-1-flat/app/service/notification.go
package service

import (
	"context"
	"log"
	"time"

	"github.com/google/uuid"

	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-1-flat/app/model"
)

type NotificationRepo interface {
	Create(ctx context.Context, n model.Notification) error
}

type NotificationService struct {
	repo NotificationRepo
}

func NewNotificationService(r NotificationRepo) *NotificationService {
	return &NotificationService{repo: r}
}

func (s *NotificationService) SendTest(ctx context.Context, userID uuid.UUID, channel, payload string) (*model.Notification, error) {
	now := time.Now().UTC()
	n := model.Notification{
		ID:      uuid.New(),
		UserID:  userID,
		Channel: channel,
		Payload: payload,
		SentAt:  &now,
	}

	log.Printf("[stub-sender] %s -> user=%s payload=%q", channel, userID, payload)

	if err := s.repo.Create(ctx, n); err != nil {
		return nil, err
	}
	return &n, nil
}
